src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -2821,43 +2832,101 @@ _mongoc_cluster_stream_for_optype (mongoc_cluster_t *cluster,
          cs, topology, optype, read_prefs, &must_use_primary, error);
 
       if (!server_id) {
-         _mongoc_bson_init_with_transient_txn_error (cs, reply);
+         if (reply) {
+            bson_init (reply);
+            _mongoc_add_transient_txn_error (cs, reply);
+         }
          RETURN (NULL);
       }
    }
 
-   /* connect or reconnect to server if necessary */
-   server_stream = _mongoc_cluster_stream_for_server (
-      cluster, server_id, true /* reconnect_ok */, cs, reply, error);
+   bson_t first_reply;
+   bson_error_t first_error = {0};
+
+   server_stream = _mongoc_cluster_stream_for_server (cluster,
+                                                      server_id,
+                                                      true /* reconnect_ok */,
+                                                      cs,
+                                                      &first_reply,
+                                                      &first_error);
+
    if (server_stream) {
       server_stream->must_use_primary = must_use_primary;
+      RETURN (server_stream);
    }
 
-   RETURN (server_stream);
+   // Important: authentication errors are also considered retryable even if
+   // they not considered a network error.
+   const bool retryable_error = _mongoc_error_is_network (&first_error) ||
+                                _mongoc_error_is_auth (&first_error);
+
+   if (is_retryable && retryable_error) {
+      bson_t retry_reply;
+      bson_error_t retry_error = {0};
+
+      server_stream =
+         _mongoc_cluster_stream_for_server (cluster,
+                                            server_id,
+                                            true /* reconnect_ok */,
+                                            cs,
+                                            &retry_reply,
+                                            &retry_error);
+
+      if (server_stream) {
+         server_stream->must_use_primary = must_use_primary;
+         server_stream->retry_attempted = true;
+         bson_destroy (&first_reply);
+         RETURN (server_stream);
+      }
+
+      if (optype != MONGOC_SS_READ) {
+         // Retryable Writes Spec: When the driver encounters a network error
+         // establishing an initial connection to a server, it MUST add a
+         // RetryableWriteError label to that error if the MongoClient
+         // performing the operation has the retryWrites configuration option
+         // set to true.
+         _mongoc_write_error_append_retryable_label (&first_reply);
+      }
+
+      bson_destroy (&retry_reply);
+   }
+
+   // Retryable Writes Spec: If the driver cannot select a server for the retry
+   // attempt [...], retrying is not possible and drivers MUST raise the
+   // original retryable error.
+   {
+      if (reply) {
+         bson_steal (reply, &first_reply);
```

Comment:
Beware: `bson_steal` seems subtly broken, and I no longer trust it. It marks the `dst` as read-only and makes `bson_destroy(dst)` never free `dst` itself (both of these because it sets `BSON_FLAG_STATIC` on `dst`). I have no idea why it does this, the function is under-documented, there's no commentary, and the git-blame is useless. I think it would be good to investigate, but not high-priority yet. It might "just work" in this case, but it has bitten me in the past (repeatedly).

Reply:
It may be the case that it has not caused any trouble so far in this PR due to all the `bson_t` in question being circumstantially stack-allocated, thus not requiring a `bson_free`. You reminded me that I had a branch a while back with a new `bson_move_to` to avoid some of the idiosyncracies I had encountered with `bson_steal`. I may need to recover and propose a PR for that later. I will make an attempt to avoid `bson_steal` for now as suggested.🤔